lib/command_factory.js

```diff
@@ -39,26 +59,72 @@ function CommandFactory(robot) {
+
+  // Maps command format to a compiled regex for that format
+  this.st2_commands_global_regex_map = {};
 }
 
-CommandFactory.prototype.getRegexForFormatString = function(format) {
-  var extra_params, regex_str, regex;
+CommandFactory.prototype.getRegexForFormatString = function(format, flags) {
+  var extra_params, regex, regex_str = format;
 
   // Note: We replace format parameters with ([\s\S]+?) and allow arbitrary
   // number of key=value pairs at the end of the string
   // Format parameters with default value {{param=value}} are allowed to
   // be skipped.
   // Note that we use "[\s\S]" instead of "." to allow multi-line values
   // and multi-line commands in general.
+
+  // Optional multiple key=value pairs
+  // ( attr_name="value"|'something'|({something})|something )*
   extra_params = '(\\s+(\\S+)\\s*=("([\\s\\S]*?)"|\'([\\s\\S]*?)\'|({[\\s\\S]*?})|(\\S+))\\s*)*';
-  regex_str = format.replace(/(\s*){{\s*\S+\s*=\s*(?:({.+?}|.+?))\s*}}(\s*)/g, '\\s*($1([\\s\\S]+?)$3)?\\s*');
-  regex_str = regex_str.replace(/\s*{{.+?}}\s*/g, '\\s*([\\s\\S]+?)\\s*');
-  regex = new RegExp('^\\s*' + regex_str + extra_params + '\\s*$', 'i');
-  return regex;
+
+  // From https://docs.stackstorm.com/chatops/aliases.html
+  // Possible inputs:
+  // Basic: "run {{cmd}} on {{hosts}}" -> /run (?<cmd>\S+) on (?<hosts>\S+)/
+  // With default: "run {{cmd}} on {{hosts=localhost}}" -> /run (?<cmd>\S+) on (?<hosts>\S+)/
+  // With JSON default: "run {{thing={'key': 'value'}}}" -> /run (?<thing>[^\s=]+).*?/
+  // Regular expressions: "(run|execute) {{cmd}}( on {{hosts=localhost}})?[!.]?"
+  // Key-value parameters: "run {{cmd}} on {{hosts}} {{key1}}={{value1}} {{key2}}={{value2}}"
+  //
+  // This function is intentionally written to be as stupid as possible. We
+  // avoid attempting parsing regex strings with regexes, as down that path
+  // lies madness.
+  //
+  // If you pass in a format string, we convert that to a regex for you, with
+  // full named capture groups and everything.
+  // But if you pass in a fully-formed regex, we completely leave that alone.
+  // Write your aliases accordingly!
+
+  if (regex_str.indexOf("{{") !== -1 && regex_str.indexOf("}}") !== -1) {
+    // {{ option=default_value }} -> (([\s\S]+?))?  // Why wrap the whole thing in an optional group?
+    regex_str = regex_str.replace(/(\s*){{\s*(\S+)\s*=\s*(?:({.+?}|.+?))\s*}}(\s*)/g, '\\s*($1(?<$2>[\\s\\S]+?)$4)?\\s*');
+
+    // {{ option }} -> ([\s\S]+?)
+    // Grab the whitespace on either side so regexes can specify whether or not the whitespace is optional
+    // regex_str = regex_str.replace(/(\s*){{\s*([\w_]+).*?\s*}}(\s*)/g, '$1(?<$2>[\\s\\S]+?)$3');
+    var parameter_re = new XRegExp('(?<before_ws>\\s*){{\\s*(?<var>[\\w_]+).*?\\s*}}(?<after_ws>\\s*)', 'g');
+    regex_str = XRegExp.replace(regex_str, parameter_re, '${before_ws}(?<${var}>[\\s\\S]+?)${after_ws}');
+
+    // Only anchor if we aren't anchored already
+    if (regex_str.charAt(0) !== '^') {
```

Comment:
I don't like how I do this, but I don't know of a way to parse Javascript's regex objects like how I did in [PR #3789](https://github.com/StackStorm/st2/pull/3789). I'm open to suggestions.